python/shotgun_data/background_task_manager.py

```diff
@@ -0,0 +1,832 @@
+
+        :returns:   The Shotgun connection for this instance in the current thread
+        """
+        if not hasattr(self._threadlocal_storage, "shotgun"):
+            # create our own private shotgun connection. This is because
+            # the shotgun API isn't threadsafe, so running multiple models in parallel
+            # (common) may result in side effects if a single connection is shared
+            self._threadlocal_storage.shotgun = sgtk.util.shotgun.create_sg_connection()
+
+            # set the maximum timeout for this connection for fluency
+            self._threadlocal_storage.shotgun.config.timeout_secs = BackgroundTaskManager._SG_CONNECTION_TIMEOUT_SECS
+
+        return self._threadlocal_storage.shotgun
+
+    def next_group_id(self):
+        """
+        Return the next available group id
+
+        :returns:    A unique group id to be used for tasks that belong to the same group.
+        """
+        group_id = self._next_group_id
+        self._next_group_id += 1
+        return group_id
+
+    def _log(self, msg):
+        """
+        Wrapper method for logging useful information to debug.
+
+        :param msg: The message to be logged.
+        """
+        self._bundle.log_debug("Task Manager: %s" % msg)
+
+    def start_processing(self):
+        """
+        Start processing of tasks
+        """
+        self._can_process_tasks = True
+        self._start_tasks()
+
+    def pause_processing(self):
+        """
+        Pause processing of tasks - any currently running tasks will
+        complete as normal.
+        """
+        self._can_process_tasks = False
+        # and just let the current threads complete...
+
+    def shut_down(self):
+        """
+        Shut down the task manager.  This clears the task queue and gracefully stops all running
+        threads.  Completion/failure of any currently running tasks will be ignored.
+        """
+        self._log("Shutting down...")
+        self._can_process_tasks = False
+
+        # clear out all storage:
+        self._running_tasks = {}
+        self._available_threads = []
+        self._pending_tasks_by_priority = {}
+        self._tasks_by_id = {}
+        self._group_task_map = {}
+        self._upstream_task_map = {}
+        self._downstream_task_map = {}
+
+        # stop all worker threads and then wait for them to complete:
+        #for thread in self._all_threads:
+        #    #thread.stop(wait_for_completion = False)
+        #    thread.quit()
+        self._log("Waiting for %d background threads to stop..." % len(self._all_threads))
+        for thread in self._all_threads:
+            thread.shut_down()
+            #thread.deleteLater()
+            #thread = None
+            #thread.stop(wait_for_completion = True)
+        self._log("Shut down successfully!")
+        self._all_threads = []
+
+    def add_task(self, cbl, priority=None, group=None, upstream_task_ids=None, args=[], **kwargs):
```

Comment:
Sorry, I should have been clearer, there was a voluntary typo in that example : `upstream_taks_ids`, missing an `s`. Because the caller made a mistake in his parameter name, it means that `add_task` didn't get the actual task ids and the caller will not get the expected behaviour because your kwargs swallowed his bad parameter and forwarded them to the task callable.